### PR TITLE
properties management

### PR DESCRIPTION
**Pull request by raffaelespazzoli**
#### What is this PR About?
article on properties management

#### How should we test or review this PR?
Standard review

#### Is there a relevant Trello card or Github issue open for this?
not aware of

#### Who would you like to review this?
cc: @redhat-cop/cant-contain-this

### CODE REVIEW

File: playbooks/app_dev/properties-management.adoc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+Your application needs to be trained to read its properties this way. If you are using Spring, there are several way of achieving this. Here is one example: +
+@Value("#{systemEnvironment['MY_EXTERNAL_ENDPOINT']}")
+private String myExternalEndpoint; +
+Things to keep in mind when adopting this approach are that if the value of a property changes, your CI/CD workflow needs to be able to change it in the pod template and if a property is added or removed, your CI/CD workflow needs to be able to manage this event appropriately. +
+In general, this approach does not scale well when you have more than a dozen of properties.
+
+=== Using One Environment Variable to Determine the Environment
+
+In this approach, we use an environment variable to determine in which environment the application is being deployed. The pod template would like the following: +
+          env: +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: wrap in code block

File: playbooks/app_dev/properties-management.adoc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            - name: ENV +
+              value: {IT|QA|PROD}
+
+The docker image of your application will need to contain all the properties for each possible environment and to be able to select the right configuration file based on the ENV environment variable. +
+Using Spring, this can be elegantly achieved using profiles. Here is an example: +
+You will have several config classes, one per environment with the following annotations: +
+@Configuration  +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: wrap in code block

File: playbooks/app_dev/properties-management.adoc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+@PropertySource("file:///<well-known-location>/application-dev.properties")  +
+@Profile("dev")
+
+Then this pod configuration to activate the right profile: +
+          env: +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: wrap in code block

File: playbooks/app_dev/properties-management.adoc [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            - name: SPRING_ACTIVE_PROFILE +
+              value: dev
+
+Notice: in this case we don’t even need the ENV variable +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: `s/Notice:/Note:/`

File: playbooks/app_dev/properties-management.adoc [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+This approach makes your CI/CD easier to implement because now you don’t have to manage changing the values of the properties or adding and removing properties. +
+This approach falls short when the number and type of your environments start to change (for example because you start provisioning environments dynamically).
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: remove line break

File: playbooks/app_dev/properties-management.adoc [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+=== ConfigMap
+
+ConfigMap is an OpenShift API object that is managed by OpenShift with the purpose of injecting application configurations. You can create a ConfigMap to contain a properties file as follows: +
+oc create configmap my-app-config --from-file=path/to/application.properties +
+This config map will be mounted as a file called `application.properties` in a directory that you can configure in your pod template. Here are the relevant sections: +
+      volumes: +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: code block

File: playbooks/app_dev/properties-management.adoc [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+              value: dev
+
+Notice: in this case we don’t even need the ENV variable +
+This approach makes your CI/CD easier to implement because now you don’t have to manage changing the values of the properties or adding and removing properties. +
+This approach falls short when the number and type of your environments start to change (for example because you start provisioning environments dynamically).
+
+=== ConfigMap
+
+ConfigMap is an OpenShift API object that is managed by OpenShift with the purpose of injecting application configurations. You can create a ConfigMap to contain a properties file as follows: +
+oc create configmap my-app-config --from-file=path/to/application.properties +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: code block

File: playbooks/app_dev/properties-management.adoc [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+          env: +
+            - name: CONFIG_LOCATION +
+              value: /etc/myapp/config +
+          volumeMounts: +
+            - name: my-app-config +
+              mountPath: /etc/myapp/config
+
+Notice that we use an environment variable to pass the location of the properties file to the application. This environment variable must match the volume mount point and never needs to change. +
+The application will need to initialize its properties by looking at the environment variables that specified the location of the properties. If you are using Spring, this can be done as follows: +
+@PropertySource("file:///${CONFIG_LOCATION}/application.properties")
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: code block

File: playbooks/app_dev/properties-management.adoc [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+This approach allows for easily sharing  properties between different applications, because they can all mount the same configmap. +
+With this approach, you have to manage an additional OpenShift API object in your CI/CD pipeline: the configmap. Someone or some process must be able to create them and update them when properties are changed, added, or removed.
+
+=== Config Store Service
+
+In this approach, a service that we will refer to as config store is servicing configurations for one or more apps. Presumably, this server can manage different environments so when an application starts it will call the service passing its identification and the environment for which it wants the configuration and the service will respond with the appropriate properties. +
+Environment and config service endpoint will have to be passed using one of the previous methods. If using environment variables it the pod template would look as follows: +
+          env: +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: code block

File: playbooks/app_dev/properties-management.adoc [hunk 10]
```diff
@@ -0,0 +1,97 @@
+```
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+---
+---
+Raffaele Spazzoli <rspazzol@redhat.com>
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+toc::[]
+
+
+== Environment Dependent Property Management Strategies for OpenShift Pipelines
+
+One of the things to think about when designing promotion pipelines for applications deployed in OpenShift is a strategy for managing environment-dependent properties. +
+When an application is promoted through environments (such a DEV, IT, QA) some of its configuration properties need to change, for example the connection string to a database or the URL of a called web service. +
+How an application expects to read its configurations is completely application-dependent. That said, over the course of several projects we have seen some patterns emerge that we have found to be successful. There is no better or worse approach - it is the responsibility of the pipeline designer to choose the best approach for a given context. +
+This blog post focuses on environment-dependent properties, but the same approaches could be potentially used for all properties, whether or not they are environment-dependent. Credential management is out of scope of this blog post.  +
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+Here are four approaches that that cover the most common use cases. 
+
+=== Using Environment Variables
+
+The properties are passed by OpenShift as environment variable. The following is a fragment of how a such a pod template (or higher structures such as replication controllers and deployment configs) would be configured: +
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+          env: +
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+            - name: MY_EXTERNAL_ENDPOINT +
+              value: http://xxx.yyy.zzz[http://xxx.yyy.zzz]
+
+Your application needs to be trained to read its properties this way. If you are using Spring, there are several way of achieving this. Here is one example: +
+@Value("#{systemEnvironment['MY_EXTERNAL_ENDPOINT']}")
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 1 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+private String myExternalEndpoint; +
+Things to keep in mind when adopting this approach are that if the value of a property changes, your CI/CD workflow needs to be able to change it in the pod template and if a property is added or removed, your CI/CD workflow needs to be able to manage this event appropriately. +
+In general, this approach does not scale well when you have more than a dozen of properties.
+
+=== Using One Environment Variable to Determine the Environment
+
+In this approach, we use an environment variable to determine in which environment the application is being deployed. The pod template would like the following: +
```
[the remark this answers is not in the document]
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 2 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 3 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 4 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 5 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 7 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 6 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 8 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 9 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

**Comment by etsauer** [comment]
@raffaelespazzoli looks better, but there are still a lot of places where we need an empty line between paragraphs. Use the docker image builder (`docker-compose up`) to see what I mean.

File: playbooks/app_dev/properties-management.adoc [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+= Environment Dependent Property Management Strategies for OpenShift Pipelines
+Raffaele Spazzoli <rspazzol@redhat.com>
+v1.0, 2017-08-28
+:toc: macro
+:toc-title:
+
+toc::[]
+
+
+
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: I would probably put an `== Overview` section here

Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+v1.0, 2017-08-28
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+
+One of the things to think about when designing promotion pipelines for applications deployed in OpenShift is a strategy for managing environment-dependent properties. +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: These paragraphs still need line separation.

File: playbooks/app_dev/properties-management.adoc [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+@PropertySource("file:///<well-known-location>/application-dev.properties")  
+@Profile("dev")
+----
+Then this pod configuration to activate the right profile: +
+----
+          env: 
+            - name: SPRING_ACTIVE_PROFILE 
+              value: dev
+----
+Note: in this case we don’t even need the ENV variable +
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: My apologies. This should have been `NOTE:` to trigger the callout formatting. Also, Can we capitalize the beginning of the sentence and remove the first line break?

File: playbooks/app_dev/properties-management.adoc [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+...           
+          env: 
+            - name: CONFIG_LOCATION 
+              value: /etc/myapp/config 
+          volumeMounts: 
+            - name: my-app-config 
+              mountPath: /etc/myapp/config
+----
+Notice that we use an environment variable to pass the location of the properties file to the application. This environment variable must match the volume mount point and never needs to change.
+The application will need to initialize its properties by looking at the environment variables that specified the location of the properties. If you are using Spring, this can be done as follows:
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: clean up paragraph spacing

File: playbooks/app_dev/properties-management.adoc [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+            - name: my-app-config 
+              mountPath: /etc/myapp/config
+----
+Notice that we use an environment variable to pass the location of the properties file to the application. This environment variable must match the volume mount point and never needs to change.
+The application will need to initialize its properties by looking at the environment variables that specified the location of the properties. If you are using Spring, this can be done as follows: 
+----
+@PropertySource("file:///${CONFIG_LOCATION}/application.properties")
+----
+This approach allows for easily sharing  properties between different applications, because they can all mount the same configmap.
+With this approach, you have to manage an additional OpenShift API object in your CI/CD pipeline: the configmap. Someone or some process must be able to create them and update them when properties are changed, added, or removed.
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: paragraph spacing

File: playbooks/app_dev/properties-management.adoc [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+== Config Store Service
+
+In this approach, a service that we will refer to as config store is servicing configurations for one or more apps. Presumably, this server can manage different environments so when an application starts it will call the service passing its identification and the environment for which it wants the configuration and the service will respond with the appropriate properties.
+Environment and config service endpoint will have to be passed using one of the previous methods. If using environment variables it the pod template would look as follows:
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: paragraph spacing

File: playbooks/app_dev/properties-management.adoc [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+----
+          env: 
+            - name: ENV 
+              value: {IT|QA|PROD} 
+            - name: CONFIG_URL 
+              value: http://myconfig.xxx.yyy[http://myconfig.xxx.yyy]
+----
+This methods is appropriate when application properties need to be reloaded at runtime. One shortcoming of this method is that your application now has a dependency on an external service (which may be down) to start. To overcome this limit, you should have defaults for all your properties so that your application can start even if the property store service is down. 
+https://github.com/Netflix/archaius/wiki[Archaius] is a NetFlix library designed to aggregate properties from different sources and ping these sources at regular interval to look for updates. It http://cloud.spring.io/spring-cloud-static/spring-cloud-netflix/1.2.4.RELEASE[integrates] with Spring Cloud.
```
Comment (etsauer) [inline playbooks/app_dev/properties-management.adoc]: paragraph spacing

File: playbooks/app_dev/properties-management.adoc (hunk 18 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 19 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 20 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 21 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 22 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

File: playbooks/app_dev/properties-management.adoc (hunk 23 above)
Comment (raffaelespazzoli) [reply playbooks/app_dev/properties-management.adoc]: done

**Comment by rdebeasi** [comment]
Looks good to me as well. Merging.

**Pull request merged**